tuplex/python/src/PythonContext.cc

```diff
@@ -884,6 +958,76 @@ namespace tuplex {
         return majType;
     }
 
+    std::vector<Partition *> PythonContext::serializeExceptions(std::vector<std::tuple<size_t, PyObject *>> exceptions, int64_t opID) {
+        std::vector<Partition *> partitions;
+
+        Schema schema(Schema::MemoryLayout::ROW, python::Type::makeTupleType({python::Type::STRING}));
+        auto driver = _context->getDriver();
+
+        Partition* partition = driver->allocWritablePartition(allocMinSize, schema, -1);
+        int64_t* rawPtr = (int64_t*)partition->lockWriteRaw();
+        *rawPtr = 0;
+        uint8_t* ptr = (uint8_t*)(rawPtr + 1);
+        size_t numBytesSerialized = 0;
+
+        // Serialize each exception to a partition using the following schema:
+        // (1) is the field containing rowNum
+        // (2) is the field containing ecCode
+        // (3) is the field containing opID
+        // (4) is the field containing pickledObjectSize
+        // (5) is the field containing pickledObject
+        for(auto &exception : exceptions) {
+            auto rowNum = std::get<0>(exception);
+            auto pyObj = std::get<1>(exception);
+            auto ecCode = ecToI64(ExceptionCode::PYTHON_PARALLELIZE);
+            auto pickledObject = python::pickleObject(python::getMainModule(), pyObj);
+            auto pickledObjectSize = pickledObject.size();
+            size_t requiredBytes = sizeof(int64_t) * 4 + pickledObjectSize;
+
+            if (partition->capacity() < numBytesSerialized + requiredBytes) {
+                partition->unlockWrite();
+                partitions.push_back(partition);
+                partition = driver->allocWritablePartition(std::max(allocMinSize, requiredBytes), schema, -1);
+                rawPtr = (int64_t *) partition->lockWriteRaw();
+                *rawPtr = 0;
+                ptr = (uint8_t * )(rawPtr + 1);
+                numBytesSerialized = 0;
+            }
+
+            *((int64_t*)(ptr)) = rowNum; ptr += sizeof(int64_t);
+            *((int64_t*)(ptr)) = ecCode; ptr += sizeof(int64_t);
+            *((int64_t*)(ptr)) = opID; ptr += sizeof(int64_t);
+            *((int64_t*)(ptr)) = pickledObjectSize; ptr += sizeof(int64_t);
+            memcpy(ptr, pickledObject.c_str(), pickledObjectSize); ptr += pickledObjectSize;
+
+            *rawPtr = *rawPtr + 1;
+            numBytesSerialized += requiredBytes;
+        }
+
+        partition->unlockWrite();
+        partitions.push_back(partition);
+
+        int eInd = 0;
+        int eOff = 0;
+        auto eNumRows = partitions[eInd]->getNumRows();
+        for (auto info : _inputPartitionInfo) {
+            auto partitionId = std::get<0>(info);
+            auto numExceptions = std::get<1>(info);
+            _inputPartitionToPythonObjectsMap[partitionId] = std::make_tuple(numExceptions, eInd, eOff);
+            while (eOff + numExceptions >= eNumRows) {
```

Comment:
I find this loop here hard to understand. Especially because eOff and eNumRows are both modified. CAn this be simplified? Or what is the logic here?

Reply:
are we sure this loop always terminates?